src/psyclone/tests/psyir/nodes/reference_test.py

```diff
@@ -598,3 +600,114 @@ def test_reference_is_write(fortran_reader):
+    assert c_ref.enters_scope(loop)
+    # For 'd' we assume a readwrite access, but since it is a local we still
+    # know that the value does not enter the scope
+    d_ref = loop.loop_body[1].arguments[0]
+    assert not d_ref.enters_scope(loop)
+
+    # 'c' value does not leave the scope (because we write after the scope)
+    assert not c_ref.escapes_scope(loop)
+    # 'a' and 'd' value does not leave the scope (because we don't use them
+    # after the scope)
+    assert not a_ref.escapes_scope(loop)
+    assert not d_ref.escapes_scope(loop)
+    # 'b' value is used after the scope
+    assert b_ref.escapes_scope(loop)
+
+    # Do the same but with global instead of local symbols
+    code = """
+    subroutine my_subroutine()
+        use other
+
+        a = 0
+        call myfunc(b)
+        c = 0
+        do i=1, 10
+            a = b + c
+            call myfunc(d)
+        enddo
+        c = b + 1
+    end subroutine"""
+    psyir = fortran_reader.psyir_from_source(code)
+    loop = psyir.walk(Loop)[0]
+    # In this case for 'd' we cannot guarantee that it is not written somewhere
+    # else (e.g. inside myfunc) and the call assumed read uses the value
+    d_ref = loop.loop_body[1].arguments[0]
+    assert d_ref.enters_scope(loop)
+
+    # 'a' and 'd' we cannot guarantee that they are used somewhere not visible
+    # (e.g. after this subroutine) so they now escape the scope
+    a_ref = loop.loop_body[0].lhs
+    assert a_ref.escapes_scope(loop)
+    d_ref = loop.loop_body[1].arguments[0]
+    assert d_ref.escapes_scope(loop)
+
+
+def test_reference_enters_scope_multiple_conditional_source(fortran_reader):
+    ''' Test that the enters_scope with two sources both inside the loop. '''
+    code = """
+    subroutine my_subroutine()
+      use other
+      INTEGER :: zice, jk, jj, ji, jpk, jpj, jpi
+
+      DO jk = 1, jpk
+         DO jj = 1, jpj
+            DO ji = 1, jpi
+               IF( tsn(ji,jj,jk) <= ztfreez(ji,jj) + 0.1d0 ) THEN; zice = 1.d0
+               ELSE                                              ; zice = 0.d0
+               ENDIF
+               zind(ji,jj,jk) = MAX (   &
+                   rnfmsk(ji,jj) * rnfmsk_z(jk),      &
+                   upsmsk(ji,jj)               ,      &
+                   zice                               &
+                   &                  ) * tmask(ji,jj,jk)
+                   zind(ji,jj,jk) = 1 - zind(ji,jj,jk)
+            END DO
+         END DO
+      END DO
+
+    end subroutine"""
+    psyir = fortran_reader.psyir_from_source(code)
+    loops = psyir.walk(Loop)
+    zice_refs = [ref for ref in psyir.walk(Reference) if ref.name == "zice"]
+
+    # For all permutations fo zice_refs and loops it should say that zice does
+    # not enter the scope
+    for zice, loop in itertools.product(zice_refs, loops):
+        assert not zice.enters_scope(loop)
+
+    # Now repeat the test but enclosing the loops in a Profile region
+    p_trans = ProfileTrans()
```

Comment:
Is this bit of the test right? What does the addition of the ProfilingRegion do?